mistletoe/markdown_renderer.py

```diff
@@ -0,0 +1,537 @@
+"""
+Markdown renderer for mistletoe.
+"""
+
+import re
+from itertools import chain
+from typing import Iterable, Sequence
+
+from mistletoe import block_token, span_token, token
+from mistletoe.base_renderer import BaseRenderer
+
+
+class BlankLine(block_token.BlockToken):
+    """
+    Blank line token. Represents a single blank line.
+    This is a leaf block token without children.
+    """
+
+    pattern = re.compile(r"\s*\n$")
+
+    def __init__(self, _):
+        pass
+
+    @classmethod
+    def start(cls, line):
+        return cls.pattern.match(line)
+
+    @classmethod
+    def read(cls, lines):
+        return [next(lines)]
+
+
+class LinkReferenceDefinition(span_token.SpanToken):
+    """
+    Link reference definition. ([label]: dest "title")
+
+    Not included in the parsing process, but called by LinkReferenceDefinitionBlock.
+
+    Attributes:
+        label (str): link label, used in link references.
+        dest (str): link target.
+        title (str): link title (default to empty).
+    """
+
+    repr_attributes = ("label", "dest", "title")
+
+    def __init__(self, match):
+        self.label, self.dest, self.title, self.dest_type, self.title_delimiter = match
+
+
+class LinkReferenceDefinitionBlock(block_token.Footnote):
+    """
+    A sequence of link reference definitions.
+    This is a container block token. Its children are link reference definition tokens.
+
+    This class inherits from Footnote and modifies the behavior of the constructor,
+    to keep the tokens in the AST.
+    """
+
+    def __new__(cls, *args, **kwargs):
+        obj = object.__new__(cls)
+        obj.__init__(*args, **kwargs)
+        return obj
+
+    def __init__(self, matches):
+        self.children = list(map(LinkReferenceDefinition, matches))
+
+
+class Fragment:
+    """
+    Markdown fragment. Used when rendering trees of span tokens into flat sequences.
+    May carry additional data in addition to the text.
+
+    Attributes:
+        text (str): markdown fragment
+    """
+
+    def __init__(self, text: str, **extras):
+        self.text = text
+        for k, v in extras.items():
+            setattr(self, k, v)
+
+
+class MarkdownRenderer(BaseRenderer):
+    """
+    Markdown renderer.
+
+    Designed to make as "clean" a roundtrip as possible, markdown -> parsing -> rendering -> markdown,
+    except for nonessential whitespace. Except when rendering with word wrapping enabled.
+
+    Includes HTMLBlocks and HTMLSpans in the parsing.
+    """
+
+    _whitespace = re.compile(r"\s+")
+
+    def __init__(self, *extras):
+        block_token.remove_token(block_token.Footnote)
+        super().__init__(
+            *chain(
+                (
+                    block_token.HTMLBlock,
+                    span_token.HTMLSpan,
+                    BlankLine,
+                    LinkReferenceDefinitionBlock,
+                ),
+                extras,
+            )
+        )
```

Comment:
It's not only readability: the way the parsers are initialized and the fact that you can't have two different parsers initialized at the same time (because they will overwrite each others' choice of block and span tokens) is a problem. But it will take a major API break to fix it -- unless there's some clever solution that I haven't thought of.

Reply:
Yeah, sure, that's the problem with using the various global variables like `_token_types` in mistletoe you are talking about, right? The "syntactic sugar refactoring" suggested above won't solve that on its own for sure.

Related: I was thinking about those renderers' constructor parameters - what about introducing explicitly named parameter `extra_tokens` instead of that current `*extras`? That would let us pass dynamically constructed options (e.g. read from the command line) to the constructors. But of course, it is an API change again. But I think we won't avoid that anyway, sooner or later...

Reply:
Yes, that's exactly it. I think renaming "extras" to "extra_tokens" and opening up for a way to specify options from the CLI sounds like a good idea.